### PR TITLE
CSP-492-AANOnboarding-ReasonToJoinTheNetwork

### PR DESCRIPTION
**Pull request by ravihejmadi**
Introduced - Controller, Model, Validator & Tests
Changes made to Web & Test projects only

### CODE REVIEW

File: src/SFA.DAS.ApprenticeAan.Web/Validators/Onboarding/ReasonToJoinTheNetworkSubmitModelValidator.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace SFA.DAS.ApprenticeAan.Web.Validators.Onboarding;
+
+public class ReasonToJoinTheNetworkSubmitModelValidator : AbstractValidator<ReasonToJoinTheNetworkSubmitModel>
+{
+    public const string NoSelectionErrorMessage = "Tell us if you are you joining because you have engaged with an ambassador in the network?";
+
+    public ReasonToJoinTheNetworkSubmitModelValidator()
+    {
+        RuleLevelCascadeMode = CascadeMode.Stop;
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Validators/Onboarding/ReasonToJoinTheNetworkSubmitModelValidator.cs:12]: Remove this cascade setting, it is not applicable

File: src/SFA.DAS.ApprenticeAan.Web/Models/Onboarding/ReasonToJoinTheNetworkViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace SFA.DAS.ApprenticeAan.Web.Models.Onboarding;
+public class ReasonToJoinTheNetworkViewModel: ReasonToJoinTheNetworkSubmitModel, IBackLink
+{
+    public string BackLink { get; set; } = null!;
+}
+
+public class ReasonToJoinTheNetworkSubmitModel
+{
+
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Models/Onboarding/ReasonToJoinTheNetworkViewModel.cs:9]: Remove empty space

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    [HttpGet]
+    public IActionResult Get()
+    {
+        var sessionModel = _sessionService.Get<OnboardingSessionModel>();
+
+        var model = new ReasonToJoinTheNetworkViewModel()
+        {
+            EngagedWithAPreviousAmbassadorInTheNetwork = sessionModel.GetProfileValue(ProfileDataId.EngagedWithAPreviousAmbassadorInTheNetwork) == null ?
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs:38]: Value from session model is retrieved twice, retrieve it once and hold it in a variable.

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        var model = new ReasonToJoinTheNetworkViewModel()
+        {
+            BackLink = Url.RouteUrl(@RouteNames.Onboarding.AreasOfInterest)!
+        };
+
+        ValidationResult result = _validator.Validate(submitmodel);
+        if (!result.IsValid)
+        {
+            sessionModel.HasEmployersApproval = null;
+            _sessionService.Set(sessionModel);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs:59]: Do not change persistence if the validation has failed.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/ReasonToJoinTheNetworkControllerTests/ReasonToJoinTheNetworkControllerGetTests.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    [MoqAutoData]
+    public void Get_ViewModelHasEmployersApproval_RestoreFromSession(
+                        [Frozen] Mock<ISessionService> sessionServiceMock,
+                        OnboardingSessionModel sessionModel,
+                        [Greedy] ReasonToJoinTheNetworkController sut)
+    {
+        sut.AddUrlHelperMock().AddUrlForRoute(RouteNames.Onboarding.AreasOfInterest);
+        sessionServiceMock.Setup(s => s.Get<OnboardingSessionModel>()).Returns(sessionModel);
+        sessionModel.ProfileData.Add(new ProfileModel { Id = ProfileDataId.EngagedWithAPreviousAmbassadorInTheNetwork, Value = "True" });
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/ReasonToJoinTheNetworkControllerTests/ReasonToJoinTheNetworkControllerGetTests.cs:57]: Need to test for `False` value as well.

**Review by DevenKShah — DISMISSED** [review]

File: src/SFA.DAS.ApprenticeAan.Web/Validators/Onboarding/ReasonToJoinTheNetworkSubmitModelValidator.cs (hunk 1 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Validators/Onboarding/ReasonToJoinTheNetworkSubmitModelValidator.cs]: Removed.

File: src/SFA.DAS.ApprenticeAan.Web/Models/Onboarding/ReasonToJoinTheNetworkViewModel.cs (hunk 2 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Models/Onboarding/ReasonToJoinTheNetworkViewModel.cs]: Removed.

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs (hunk 4 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs]: Removed.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/ReasonToJoinTheNetworkControllerTests/ReasonToJoinTheNetworkControllerGetTests.cs (hunk 5 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/ReasonToJoinTheNetworkControllerTests/ReasonToJoinTheNetworkControllerGetTests.cs:72]: Added Test for False

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs (hunk 3 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/ReasonToJoinTheNetworkController.cs]: Moved Get operation into a variable.

**Review by DevenKShah — DISMISSED** [review]

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**